content/blog/2020-03-dplyr-1-0-0-and-vctrs/index.Rmarkdown

```diff
@@ -0,0 +1,169 @@
+
+In this first example, we attempt to bind two data frames together where the columns have incompatible types: double and character.
+
+```{r, error = TRUE}
+df1 <- tibble(a = 1, b = 1)
+df2 <- tibble(a = 2, b = "a")
+bind_rows(df1, df2)
+```
+Note the components of the error message:
+
+* "No common type" means that vctrs can't find a vector type that can
+  represent both double and character values. 
+  
+* vctrs error messages always puts the "type" of the variable in `<>`.
+  I'm using type informally here (although it does have a well-formed
+  definition); for many simple cases it's the same as the class.
+  
+* `bind_rows()` doesn't have named arguments so vctrs uses `..1` and 
+  `..2` to refer to the first and second arguments. You can tell the
+  problem is with the `b` column.
+  
+Where possible we attempt to give you more information to solve the problem. For example, if your call to `summarise()` or `mutate()` returns incompatible types, we'll tell you which groups have the problem:
+
+```{r, error = TRUE}
+df <- tibble(g = c(1, 2))
+df %>% 
+  group_by(g) %>% 
+  mutate(y = if (g == 1) "a" else 1)
+```
+Writing good error messages is hard, and we've spent a lot of time trying to make them informative. 
+
+If you're not sure where the errors are coming from, learning how to use the traceback (either `traceback()` or `rlang::last_error()`) will be helpful. I'd highliy recommend Jenny Bryan's rstudio::conf keynote on debugging: [Object of type 'closure' is not subsettable](https://resources.rstudio.com/rstudio-conf-2020/object-of-type-closure-is-not-subsettable-jenny-bryan)
+
+## Key changes
+
+Finally, the switch to vctrs causes two major changes in dplyr 1.0.0:
+
+*   When combining factors with different level sets, dplyr previously 
+    converted to a character vector with a warning. As of 1.0.0, dplyr will
+    create a factor with the union of the individual levels.
+  
+    ```{r}
+    vec_c(factor("x"), factor("y"))
+    ```
+  
+*   When combining a factor and a character, dplyr previously warned about
+    creating a character vector. It now silently creates a character vector.
+
+    ```{r}
+    vec_c("x", factor("y"))
+    ```
+
+These changes are motivated primarily by pragmatism than by theory. Strictly speaking one should probably consider `factor("red")` and `factor("male")` to be incompatible, but this level of strictness causes much pain because character vectors are usually interchangeably with factors.
+
+Also note that dplyr continues to be stricter than base R when it comes to character conversions: 
+
+```{r, error = TRUE}
+vec_c(1, "a")
+```
```

Comment:
Here's where I'd point out how one might work around this in the event that one _did_ want to combine the two (either through `as.character()` or allowing a lossy cast).